File: demo/src/main/java/com/umut/petexpandable/MusicExpandableViewAdapter.java
package com.umut.petexpandable;

import android.support.annotation.NonNull;
import android.support.v7.widget.AppCompatImageView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.umut.expandablrecyclerview.adapter.ChildCoordinate;
import com.umut.expandablrecyclerview.adapter.ExpandableViewAdapter;
import com.umut.expandablrecyclerview.adapter.holder.ChildViewHolder;
import com.umut.expandablrecyclerview.adapter.holder.ParentViewHolder;
import com.umut.petexpandable.model.MainStreamGenre;
import com.umut.petexpandable.model.SubGenre;

import java.util.List;
import java.util.Map;

public class MusicExpandableViewAdapter extends ExpandableViewAdapter {

    @NonNull
    private final List<MainStreamGenre> mainStreamGenres;
    @NonNull
    private final Map<MainStreamGenre, List<SubGenre>> subGenres;

    public MusicExpandableViewAdapter(@NonNull List<MainStreamGenre> mainStreamGenres,
                                      @NonNull Map<MainStreamGenre, List<SubGenre>> subGenres) {
        super(new SampleDataIndexProvider(mainStreamGenres, subGenres));
        this.mainStreamGenres = mainStreamGenres;
        this.subGenres = subGenres;
    }

    @Override
    protected MainStreamViewHolder createParentViewHolder(@NonNull ViewGroup container) {
        View view = LayoutInflater.from(container.getContext()).inflate(R.layout.item_main_stream,
                container, false);
        return new MainStreamViewHolder(view);
    }

    @Override
    protected SubGenreViewHolder createChildViewHolder(@NonNull ViewGroup container) {
        View view = LayoutInflater.from(container.getContext()).inflate(R.layout.item_sub_genre,
                container, false);
        return new SubGenreViewHolder(view);
    }

    private class MainStreamViewHolder extends ParentViewHolder {

        private final TextView nameTextView;
        private final AppCompatImageView expandIndicatorImageView;

        public MainStreamViewHolder(View itemView) {
            super(itemView);
            nameTextView = (TextView) itemView.findViewById(R.id.text_view_main_stream_genre_name);
            expandIndicatorImageView = (AppCompatImageView) itemView.findViewById(R.id.image_view_expand_indicator);
        }

        @Override
        public void bind(int coordinate, int state) {
            itemView.setOnClickListener((view) -> {
                toggle(coordinate);
            });
            nameTextView.setText(mainStreamGenres.get(coordinate).name);
            setState(state == State.EXPANDED);
        }

        private void setState(boolean expand) {
            if (!expand) {
                expandIndicatorImageView.setImageResource(R.drawable.ic_arrow_down);
            } else {
                expandIndicatorImageView.setImageResource(R.drawable.ic_arrow_up);
            }
        }

    }

    private class SubGenreViewHolder extends ChildViewHolder {


        private TextView nameTextView;

        public SubGenreViewHolder(View itemView) {
            super(itemView);
            nameTextView = (TextView) itemView.findViewById(R.id.text_view_sub_genre_name);
        }

        @Override
        public void bind(@NonNull ChildCoordinate coordinate) {
            nameTextView.setText(
                    subGenres.get(mainStreamGenres.get(coordinate.parentIndex))
                            .get(coordinate.childRelativeIndex)
                            .name);
        }

    }
}

File: demo/src/main/java/com/umut/petexpandable/MusicExpandableViewAdapter.kt
package com.umut.petexpandable

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import com.umut.expandablerecyclerview.ExpandableRecyclerViewAdapter
import com.umut.expandablerecyclerview.holder.ChildViewHolder
import com.umut.expandablerecyclerview.holder.ParentViewHolder
import com.umut.expandablerecyclerview.item.ParentViewItem
import com.umut.expandablerecyclerview.position.ItemPosition
import com.umut.petexpandable.model.MainStreamGenre
import com.umut.petexpandable.model.MainStreamGenre.Companion.ROCK_VIEW_TYPE
import com.umut.petexpandable.model.SubGenre

class MusicExpandableViewAdapter(
    private val mainStreamGenres: List<MainStreamGenre>,
    private val subGenres: Map<MainStreamGenre, List<SubGenre>>
) : ExpandableRecyclerViewAdapter<SubGenre, MainStreamGenre>() {

    override fun createParentViewHolder(container: ViewGroup, viewType: Int): MainStreamViewHolder {
        if (viewType == ROCK_VIEW_TYPE) {
            val view = LayoutInflater.from(container.context).inflate(
                R.layout.item_main_stream_red,
                container, false
            )
            return MainStreamViewHolder(view)
        }
        val view = LayoutInflater.from(container.context).inflate(
            R.layout.item_main_stream,
            container, false
        )
        return MainStreamViewHolder(view)
    }

    override fun createChildViewHolder(container: ViewGroup, viewType: Int): SubGenreViewHolder {
        if (viewType == ROCK_VIEW_TYPE) {
            val view = LayoutInflater.from(container.context).inflate(
                R.layout.item_sub_genre_red,
                container, false
            )
            return SubGenreViewHolder(view)
        }
        val view = LayoutInflater.from(container.context).inflate(
            R.layout.item_sub_genre,
            container, false
        )
        return SubGenreViewHolder(view)
    }

    inner class MainStreamViewHolder(private val view: View) :
        ParentViewHolder<MainStreamGenre>(view) {

        override fun bind(position: ItemPosition, item: MainStreamGenre) {
            view.setOnClickListener {
                if (!item.isExpanded()) {
                    expand(position.childIndex)
                } else {
                    collapse(position.childIndex)
                }
            }
            nameTextView.text = mainStreamGenres[position.childIndex].name
            setState(item.isExpanded())
        }

        override fun onToggle(parentItem: ParentViewItem) {
            setState(parentItem.isExpanded())
        }

        private val nameTextView: TextView =
            itemView.findViewById(R.id.text_view_main_stream_genre_name)
        private val expandIndicatorImageView: AppCompatImageView =
            itemView.findViewById(R.id.image_view_expand_indicator)


        private fun setState(expand: Boolean) {
            if (!expand) {
                expandIndicatorImageView.setImageResource(R.drawable.ic_arrow_down)
            } else {
                expandIndicatorImageView.setImageResource(R.drawable.ic_arrow_up)
            }
        }

    }

    inner class SubGenreViewHolder internal constructor(itemView: View) :
        ChildViewHolder<SubGenre>(itemView) {
        override fun bindChild(
            nestedPosition: ItemPosition,
            item: SubGenre
        ) {
            nameTextView.text = item.name
        }


        private val nameTextView: TextView = itemView.findViewById(R.id.text_view_sub_genre_name)

    }
}
